modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java

```diff
@@ -344,4 +364,24 @@ private void proceedWithMaintenance() {
     @Nullable @Override public MaintenanceTask requestedTask(String maintenanceTaskName) {
         return requestedTasks.get(maintenanceTaskName);
     }
+
+    /** Starts a thread which will shut down the node when all maintenance tasks would be completed */
+    private void startShutdownThread() {
+        String igniteInstanceName = ctx.config().getIgniteInstanceName();
+
+        IgniteThread t = new IgniteThread(new GridWorker(igniteInstanceName, "maintenance-shutdown-thread", log) {
+            @Override protected void body() {
+                while (true) {
+                    if (activeTasks.isEmpty() && Ignition.state(igniteInstanceName) == IgniteState.STARTED) {
+                        G.stop(igniteInstanceName, false);
+                        return;
+                    }
+
+                    LockSupport.parkNanos(100_000_000);
```

Comment:
Why are you using `LockSupport#parkNanos` instead of `Thread#sleep`?
Maybe periodically log the number of unfinished tasks?
Can we log when the tasks are over and we go to shut down the node?

Reply:
done